### ISSUE TITLE
Unit tests fail seemingly due to an error in the test script 

### ISSUE THREAD

**Issue by gwselke** [reneeb/Map-Tube-Frankfurt#8]
The standard unit test fails with:
`t/map.t ... 1/3 Map::Tube::Plugin::Formatter::to_string(): ERROR: Missing plugin Map::Tube::Plugin::Formatter. (status: 122) file t/map.t on line 254`
This is the offending test script line:
`diag $map->to_string( $map->get_shortest_route( 'Riedstadt-Goddelau', 'Hauptbahnhof' ) );`
`to_string( )` is used here as a method of the map object, which in fact it isn't.
The test actually passes with this modified statement:
`diag $map->get_shortest_route( 'Riedstadt-Goddelau', 'Hauptbahnhof' );`
or with this statement:
`diag $map->get_shortest_route( 'Riedstadt-Goddelau', 'Hauptbahnhof' )->to_string( );`

**Comment by paultcochrane** [comment]
I notice in the output the message

```
ERROR: Missing plugin Map::Tube::Plugin::Formatter
```

have you tried installing `Map::Tube::Plugin::Formatter` and seeing if the tests then pass?  It's my guess that because this formatter isn't available, that it's causing the issue you're seeing.

**Comment by gwselke** [comment]
Thank you for checking on this -- but no, I do have Map::Tube::Plugin::Formatter v0.17 installed, and it works indeed, in other contexts. In particular, it works (as indicated above) for the result of `get_shortest_route( )`.
I don't quite get it how this specific error message comes about. However, the problem seems to be rather that this specific plugin provides `to_string( )` functionality just for Nodes, Lines, and Routes, but not for complete Maps. (And quite understandably so, because the stringification of the whole map data is generally large and probably not very useful.)
Internally, the `to_string( )` method just calls the object's `as_string( )` method, which is provided by Node, Line, and Route objects, but (AFAICT) not by Map objects. So the delegation fails when called on Map objects.

**Comment by paultcochrane** [comment]
Well, it was an attempt at an explanation :-)  Weird that the error message appears though.  Interestingly enough, it *doesn't* appear on my system (Debian Linux, perl 32.1, `Map::Tube` version 3.89, `Map::Tube::Plugin::Formatter` version 0.17).  I'm wondering if there's a version issue going on here. What versions do you have for `Map::Tube` and `Map::Tube::Plugin::Formatter`?

What made me look at the issues in this project in the first place was a different failure, one that looks like a UTF-8 handling issue in `Map::Tube`, but I've not been able to narrow it down completely yet. Are you seeing any other test failures on your system?

**Comment by gwselke** [comment]
Yes, and I sure DO appreciate your help, Paul! I'm running Perl 5.40.0 (newer than yours), Map::Tube 3.89, Map::Tube::Plugin::Formatter 0.17 (both same as yours). If you look at the CPAN testers' matrix (http://matrix.cpantesters.org/?dist=Map-Tube-Frankfurt+0.04&r=1), you'll see plenty of FAILs including your version of Perl (except, strangely, on even older Perls -- but these are tests that were run 5 years ago, so probably not with the latest Map::Tube). -- So it seems you are not getting an error where CPAN testers do get FAILs. (Lucky you :-) )

As indicated above, I think the problem is that the to_string( ) method provided by M::T::P::Formatter is *not* available for a map object. Looking at the definition of to_string( ), it was never meant to to take an argument -- that's also the message from the M::T::P::Formatter POD. But the offending line in the M::T::Frankfurt test does exactly that: call to_string( ) on a map object and passing an argument to that call.

**Comment by gwselke** [comment]
As for your question, Paul: no, otherwise I do not get any other test failures on M::T::Frankfurt. Or which module are you referring to that gives you trouble, and how does it manifest?

**Comment by paultcochrane** [comment]
Thanks for the feedback, it looks like it's not a version issue (one possible cause for the differences we're seeing out of the way!).

The failures I'm getting are due to UTF-8 string handling.  If you look at e.g. this test failure: http://www.cpantesters.org/cpan/report/e206df38-cb74-11ef-af57-c7cfaf431ec4, you'll not only see the error that you mention in this issue

```
Map::Tube::Plugin::Formatter::to_string(): ERROR: Missing plugin Map::Tube::Plugin::Formatter. (status: 122) file t/map.t on line 25
```

but also this error:

```
Map::Tube::get_shortest_route(): ERROR: Route not found from [Hochheim (Main)] to [FlÃ¶rsheim (Main)]. (status: 133) file /home/sand/src/perl/repoperls/installed-perls/host/k93jammy/v5.41.7/5674/lib/site_perl/5.41.7/Test/Map/Tube.pm on line 223

#   Failed test at t/map.t line 20.
#          got: 0
#     expected: 1
```

which is due to the station name "Flörsheim (Main)" not being recognised as a valid station name. I've managed to work out that there's a point in the code where the JSON read in from the input file gets converted into UTF-8 for what *looks like* the second time, causing the umlauts to be removed from the string and thus when the route names are used later (and they contain umlauts), they don't correspond to any station name that `Map::Tube` knows about.  I'm still working out how to formulate the bug report, but as soon as I've written up what I know, I'll link it back to this issue here.

Interestingly enough it's not just umlauts causing problems: the "sharp s" (ß) is being converted internally by `uc()` into ß in one case and SS in another case (the sharp s can also be written as double-s in German).  Both forms are technically true in terms of the German language, however for the purposes of `Map::Tube` since both forms turn up they cause comparisons to fail and hence the station name checking code doesn't recognise the station name from the route list as being valid and hence the tests fail.  This issue is related to the one about umlauts, but is slightly different.  I've not yet worked out how to explain this issue clearly in a bug report yet, but that'll happen eventually.

**Comment by gwselke** [comment]
According to what I see, the problem described by Paul is caused by a bug in Map::Tube (not in Map::Tube::Frankfurt). I suggest to take the umlaut problem there.

The initially described problem with to_string( ), however, does remain. It looks like a rather easy-to-fix problem in the test suite, as indicated above.
